### PR TITLE
imx7: pm: generate struct imx7_pm_info offsets automatically

### PR DESCRIPTION
**Pull request by jforissier**

### CODE REVIEW

**Comment by jforissier** [comment]
@MrVan can you please review the 3rd commit in particular? ("plat-imx: pm: generate struct imx7_pm_info offsets automatically"). Thanks!

File: core/arch/arm/plat-imx/sub.mk [hunk 1]
```diff
@@ -8,6 +8,8 @@ ifeq ($(CFG_PSCI_ARM32),y)
 srcs-y += pm/psci.c pm/gpcv2.c
 srcs-$(CFG_MX7) += pm/pm-imx7.c pm/psci-suspend-imx7.S pm/imx7_suspend.c
 $(call force,CFG_PM_ARM32,y)
+asm-defines-file := $(sub-dir)/imx_pm_asm_defines.c
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-imx/sub.mk]: This somewhat hackish.

Comment (jforissier) [reply core/arch/arm/plat-imx/sub.mk]: What is hackish exactly? The use of `$(sub-dir)` when other statements such as `srcs-y` use relative paths? Or the fact that simply including `mk/gen-asm-defines.mk` will generate a file? (would you rather recommend that we should call a function instead?)

**Comment by jforissier** [comment]
Update: moved the `DEFINE` and `DEFINES` macros to their own header file to avoid duplication and checkpatch warnings.

File: core/arch/arm/plat-imx/sub.mk (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/plat-imx/sub.mk]: It's a bit of both, `asm-defines-y += imx_pm_asm_defines.c` without any special include would harmonize better with the rest.

However, I don't think it's much of an issue as we're likely to have at most one of these files per platform directory.

Comment (jforissier) [reply core/arch/arm/plat-imx/sub.mk]: I also prefer you suggestion. It makes things much cleaner. Patches updated and squashed.

**Comment by MrVan** [comment]
@jforissier `Tested-by: Peng Fan <peng.fan@nxp.com>`, for the 3rd commit, `Reviewed-and-Tested-by: Peng Fan <peng.fan@nxp.com>`

**Comment by jforissier** [comment]
@jenswi-linaro please let me know if your Ack still applies, thanks.

File: core/include/gen-asm-defines.h [hunk 2]
```diff
@@ -0,0 +1,7 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/* Copyright (c) 2018, Linaro Limited */
```
Comment (etienne-lms) [inline core/include/gen-asm-defines.h]: `#ifndef __GEN_ASM_DEFINES_H `... ?

File: mk/compile.mk [hunk 3]
```diff
@@ -223,8 +223,4 @@ define gen-asm-defines-file
 $(call _gen-asm-defines-file,$1,$2,$(dir $2).$(notdir $(2:.h=.s)))
 endef
 
-ifneq ($(asm-defines-file),)
-h-file-$(asm-defines-file) := $(out-dir)/$(sm)/include/generated/$(basename $(notdir $(asm-defines-file))).h
-$(eval $(call gen-asm-defines-file,$(asm-defines-file),$(h-file-$(asm-defines-file))))
-asm-defines-file :=
-endif
+$(foreach f,$(asm-defines-files),$(eval $(call gen-asm-defines-file,$(f),$(out-dir)/$(sm)/include/generated/$(basename $(notdir $(f))).h)))
```
Comment (etienne-lms) [inline mk/compile.mk]: maybe split over 2 or 3 lines.

File: core/include/gen-asm-defines.h (hunk 2 above)
Comment (jforissier) [reply core/include/gen-asm-defines.h]: Oops, will fix.

File: mk/compile.mk (hunk 3 above)
Comment (jforissier) [reply mk/compile.mk]: TBH I tried but you don't want to introduce spaces at some places or the command will fail so I did not bother ;-)

Comment (etienne-lms) [reply mk/compile.mk]: If not handy, forget it. It's about being easier to understand/maintain.

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
@etienne-lms do you want to provide a tag?
I'm ignoring the CI failures:
- HAB is broken due to an upstream l-loader issue: https://github.com/96boards-hikey/l-loader/pull/15
- Travis fails because we're using GCC 4.9 and that does not work since manifest.git and build.git have moved to buildroot.

File: core/include/gen-asm-defines.h [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+#ifndef GEN_ASM_DEFINES_H
```
Comment (etienne-lms) [inline core/include/gen-asm-defines.h]: missing `#define GEN_ASM_DEFINES_H`

File: scripts/checkpatch_inc.sh [hunk 5]
```diff
@@ -2,7 +2,8 @@
 
 CHECKPATCH="${CHECKPATCH:-checkpatch.pl}"
 # checkpatch.pl will ignore the following paths
-CHECKPATCH_IGNORE=$(echo core/lib/lib{fdt,tomcrypt} core/lib/zlib \
+CHECKPATCH_IGNORE=$(echo core/include/gen-asm-defines.h \
```
Comment (etienne-lms) [inline scripts/checkpatch_inc.sh]: minor: can change to this for later additions:
```shell
+CHECKPATCH_IGNORE=$(echo \
+		core/include/gen-asm-defines.h \ 
+		core/lib/lib{fdt,tomcrypt} core/lib/zlib \ 
```

**Comment by etienne-lms** [comment]
with header file fixed: `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Pull request merged**